### ISSUE TITLE
strings: formatting makes strings barely translatable

### ISSUE THREAD

**Issue by silopolis** [excalidraw/excalidraw#6351]
Hello,

As reported [on Crowdin](https://crowdin.com/translate/excalidraw/all/en-fr?filter=basic&value=0#h=eb95fdc5ea1ffc4140a859d6ff7dd9d71635043) by @bancek, exploded sentences make them barely translatable.

https://github.com/excalidraw/excalidraw/blob/023313e92f7a0173c06780d4468bc314dcd816ac/src/components/__snapshots__/App.test.tsx.snap#L8-L59

**Comment by bancek** [comment]
Introduced in #6336

**Comment by ad1992** [comment]
I completely agree. We need to update our translation algo to support these cases via interpolation instead so that sentences could be used with links

**Comment by bancek** [comment]
At [Koofr](https://koofr.eu/) we use a custom `React` component and i18n extraction scripts to handle more complex translations mixed with other potentially nested React components.

Example:

```jsx
<Interpolate
  i18nKey={
    "Provider {{provider}} is already connected. To connect another account " +
    "please {{logoutLinkStart}}logout from account {{accountName}}{{logoutLinkEnd}} first."
  }
  provider={<strong>{providerName}</strong>}
  logoutLink={(el) => <a href={logoutUrl}>{el}</a>}
  accountName={accountName}
/>
```

The Slovenian translation is then just:

```
Storitev {{provider}} je že povezana. Za povezavo dodatnega
računa se najprej {{logoutLinkStart}}odjavite iz računa {{accountName}}{{logoutLinkEnd}}.
```

**Comment by ad1992** [comment]
> At [Koofr](https://koofr.eu/) we use a custom `React` component and i18n extraction scripts to handle more complex translations mixed with other potentially nested React components.
> 
> Example:
> 
> ```js
> <Interpolate
>   i18nKey={
>     "Provider {{provider}} is already connected. To connect another account " +
>     "please {{logoutLinkStart}}logout from account {{accountName}}{{logoutLinkEnd}} first."
>   }
>   provider={<strong>{providerName}</strong>}
>   logoutLink={(el) => <a href={logoutUrl}>{el}</a>}
>   accountName={accountName}
> />
> ```
> 
> The Slovenian translation is then just:
> 
> ```
> Storitev {{provider}} je že povezana. Za povezavo dodatnega
> računa se najprej {{logoutLinkStart}}odjavite iz računa {{accountName}}{{logoutLinkEnd}}.
> ```
> 
> This was inspired by `i18next-react`: https://react.i18next.com/latest/trans-component

I see, the same library I pinged above :D , yeah we would have to do something similar to handle complex translations

**Comment by Contextualist** [comment]
👍 Glad that this is brought up again, the proposed solution looks much cleaner than my previous manual implementation in #4268. This will also benefit the i18n strings on submitting a library (as shown in my previous PR).

**Comment by ad1992** [comment]
> I'd prefer not to migrate over to react-i18next just because of the Trans component.

@dwelle can you elaborate on this ? Trans comp will be only needed when we have JSX mixed with styling.

**Comment by dwelle** [comment]
> > I'd prefer not to migrate over to react-i18next just because of the Trans component.
> 
> @dwelle can you elaborate on this ? Trans comp will be only needed when we have JSX mixed with styling.

Not sure I understand. We'll need something like the Trans component regardless of style or other attributes, if we want to interpolate JSX elements.

**Comment by ad1992** [comment]
> > > I'd prefer not to migrate over to react-i18next just because of the Trans component.
> > 
> > 
> > @dwelle can you elaborate on this ? Trans comp will be only needed when we have JSX mixed with styling.
> 
> Not sure I understand. We'll need something like the Trans component regardless of style or other attributes, if we want to interpolate JSX elements.

No `Trans` will only be needed when we want to interpolate styling in JSX eg `<p><strong>Hello</strong>World</p>`, otherwise Trans will not be needed so this would be rarely used.
Right now those are the cases where i18n is breaking for Excal so in those case we need trans comp.
But to make sure we are on same page may be you can give an example of any particular case you are talking about ?

**Comment by ad1992** [comment]
> > > > I'd prefer not to migrate over to react-i18next just because of the Trans component.
> > > 
> > > 
> > > @dwelle can you elaborate on this ? Trans comp will be only needed when we have JSX mixed with styling.
> > 
> > 
> > Not sure I understand. We'll need something like the Trans component regardless of style or other attributes, if we want to interpolate JSX elements.
> 
> No `Trans` will only be needed when we want to interpolate styling in JSX eg `<p>\<strong\>Hello\</strong\>World</p>`, otherwise Trans will not be needed so this would be rarely used. Right now those are the cases where i18n is breaking for Excal so in those case we need trans comp. But to make sure we are on same page may be you can give an example of any particular case you are talking about ?

To Clarify further, for any interpolation in JSX, Trans comp would be right choice as we end up splitting strings which affects the language translations.
So for the links as well (I think they supported it via strings interpolation but can't find it in docs anymore so probably it was removed or I misremember it) and trans comp seems to be the right choice for links as well as ultimately we are splitting strings.
So to summarise I think we can check if there are some other libraries which are light weight than `react-i18next` (7kb gzipped) and decide on next steps.

**Comment by bancek** [comment]
@ad1992 @dwelle I've added a PoC PR https://github.com/excalidraw/excalidraw/pull/6534
